File: src/org/codepulsar/ast/expressions/Binary.java
package org.codepulsar.ast.expressions;

import org.codepulsar.ast.Expression;

public class Binary extends Expression {
    private final Expression left;
    private final String operator;
    private final Expression right;

    private final int line;

    public Binary(Expression left, String operator, Expression right, int line) {
        this.left = left;
        this.operator = operator;
        this.right = right;
        this.line = line;
    }

    public <R> R accept(Visitor<R> visitor) {
        return visitor.visitBinaryExpression(this);
    }

    public Expression getLeft() {
        return this.left;
    }

    public String getOperator() {
        return this.operator;
    }

    public Expression getRight() {
        return this.right;
    }

    public int getLine() {
        return this.line;
    }
}

File: src/codepulsar/ast/expression/Binary.cpp
#include "Binary.h"


Pulsar::Binary::Binary(Expression* left, std::string operatorType, Expression* right, int line) {
    this->left = left;
    this->operatorType = operatorType;
    this->right = right;
    this->line = line;
}

std::any Pulsar::Binary::accept(ExprVisitor& visitor) {
    return visitor.visitBinaryExpression(this);
}

Pulsar::Expression* Pulsar::Binary::getLeft() {
    return this->left;
}

std::string Pulsar::Binary::getOperator() {
    return this->operatorType;
}

Pulsar::Expression* Pulsar::Binary::getRight() {
    return this->right;
}

int Pulsar::Binary::getLine() {
    return this->line;
}
